scan all rows of CSV to guess file types

cjrace opened the pull request:
## Overview of changes

Fixes #2. Forces duckdb to read the whole CSV before guessing a column type.

I wanted to do something fancier, by extracting indicators and then passing them into the options in duckdb's auto detection to force all indicators to VARCHAR, however that option only takes one column - https://duckdb.org/docs/stable/data/csv/auto_detection. So the easiest approach was to force scanning the full file. I expect a mild performance hit, but still likely a lot faster than the previous reading of CSVs in the Shiny app.

## Why are these changes being made?

We have some awkward columns that are numeric, but with occasional character symbols. This was tripping up the function trying force the likes of 'x' to a double.

Moving the metadata to use read.csv, as a) it's faster and b) we have code on the metadata file that triggers fallbacks away from duckplyr and those messages can build up in the console, so read.csv is less noisy for how we scan the metadata.

Explicitly referencing some data masking to bat away linting notes.

## Checklist

- [x] I have tested these changes locally using `shinytest2::test_app()`<!-- replace with your specific automated test command if different -->
- [x] I have updated the documentation
- [x] I have added or updated automated tests for these changes

## Reviewer instructions

Test with the problematic file yourself to make sure, but otherwise if the automated tests pass I'm pretty confident in these changes so not a lot of testing needed.

cjrace committed:
scan all rows of CSV to guess file types

rmbielby reviewed (commented):
Thanks for picking this up. Couple of thoughts on this:
1. Do we definitely need to pre-determine the columns? I'd been wondering whether we can just tell it to read everything in as a string? As in are there many checks that care whether the data's numeric or not? Any that handle indicator entries will have to be able to deal with x, c, z, etc, so anything that does care is going to have some processing to handle those regardless, so is it easier if everything's just a string from the point of read-in?
2. I changed the meta read in away from read.csv as we were having some weird issues whereby it seemed to identify non-allowed characters at the start of the file and therefore added X.... to the first column name, making a few files fail the meta col names column name check. Inspecting the files pre-upload didn't show any signs of there being a special character, so I figured it was something to do with how the Azure copy of the file into the screener api app container and read.csv were interacting. I switched to the duckdb read in and the issue seems to have gone away since I did that, so if you switch back to read.csv, we're going to need to find out why it thinks there are special characters at the start of the file.

<img width="2256" height="1504" alt="image" src="https://github.com/user-attachments/assets/fc70acd1-63b4-4465-b14a-07bc89c16c53" />

cjrace committed:
revert back to duckdb for metadata files but immediately materialise

cjrace commented:
> Thanks for picking this up. Couple of thoughts on this:
> 
> 1. Do we definitely need to pre-determine the columns? I'd been wondering whether we can just tell it to read everything in as a string? As in are there many checks that care whether the data's numeric or not? Any that handle indicator entries will have to be able to deal with x, c, z, etc, so anything that does care is going to have some processing to handle those regardless, so is it easier if everything's just a string from the point of read-in?
> 2. I changed the meta read in away from read.csv as we were having some weird issues whereby it seemed to identify non-allowed characters at the start of the file and therefore added X.... to the first column name, making a few files fail the meta col names column name check. Inspecting the files pre-upload didn't show any signs of there being a special character, so I figured it was something to do with how the Azure copy of the file into the screener api app container and read.csv were interacting. I switched to the duckdb read in and the issue seems to have gone away since I did that, so if you switch back to read.csv, we're going to need to find out why it thinks there are special characters at the start of the file.
> 
> <img alt="image" width="2000" height="1504" src="https://private-user-images.githubusercontent.com/230859/527997863-fc70acd1-63b4-4465-b14a-07bc89c16c53.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjYxMDI3OTYsIm5iZiI6MTc2NjEwMjQ5NiwicGF0aCI6Ii8yMzA4NTkvNTI3OTk3ODYzLWZjNzBhY2QxLTYzYjQtNDQ2NS1iMTRhLTA3YmM4OWMxNmM1My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIxOVQwMDAxMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDI0MzRkMWQ5YzliNWMyYjZkMTAzOWQ2ZjRhYzgxMmNkYWZkNTAyZTcxY2MxYTg0ZGY4ZTM2NjQ4MDAwYTBlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.q5PEoK64MjjDFNFAhvdPgYUeBrTHMxELlF3YF6jn1ZI">

Good to know - didn't realise about the `read.csv()` issue! Have reverted back to duckdb, but immediately materialising once read in. Minimal performance impact but hopefully keeps the safer method for reading in (found on other scripts recently that duckdb is just much better out of the box at handling types without issue).

For 1. I think we should - most reading functions do this, and it doesn't seem to have much of a performance hit even on these 'bigger' files, fairly sure it's still faster than the old way in the Shiny app, so I think we keep it as this for now rather than forcing to all characters.

Think this is ready for re-review @rmbielby

rmbielby committed:
Merge branch 'main' into 2-fix-csv-type-detection
